src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.State.cs

```diff
@@ -677,11 +677,16 @@ public bool DeclarationsAccessibleWithoutAlias(int referencedAssemblyIndex)
             return aliases.Length == 0 || aliases.IndexOf(MetadataReferenceProperties.GlobalAlias, StringComparer.Ordinal) >= 0;
         }
 
-        internal override IEnumerable<KeyValuePair<AssemblyIdentity, PortableExecutableReference>> GetImplicitlyResolvedAssemblyReferences()
+        internal override IEnumerable<(PEAssembly, PortableExecutableReference)> GetImplicitlyResolvedAssemblyReferences()
         {
             foreach (PortableExecutableReference reference in ImplicitReferences)
             {
-                yield return KeyValuePairUtil.Create(ReferencedAssemblies[ReferencedAssembliesMap[reference]].Identity, reference);
+                var assembly = ReferencedAssemblies[ReferencedAssembliesMap[reference]];
+
+                // implicitly resolved references are PE assemblies
+                Debug.Assert(assembly.Assembly != null);
+
+                yield return (assembly.Assembly, reference);
```

Comment:
>assembly.Assembly [](start = 30, length = 17)

Is this where the new ```IAssemblySymbolInternal.Assembly``` API is used? Can we get the same information from PortableExecutableReference instead?

Reply:
No, we can't. We need to reuse the AssemblyMetadata object that we already used before. PEReference may return a different instance if we read it again.

Reply:
> PEReference may return a different instance if we read it again.

I am assuming you are talking about a copy object. Why is it important to get the same instance? 

---
In reply to: [322395639](https://github.com/dotnet/roslyn/pull/38550#discussion_r322395639) [](ancestors = 322395639)